tools/ci_build/github/azure-pipelines/templates/react-native-ci.yml

```diff
@@ -20,302 +20,338 @@ parameters:
-
-  - template: android-dump-logs-from-steps.yml
-    parameters:
-      steps:
-      - task: Gradle@3
-        inputs:
-          gradleWrapperFile: '$(Build.SourcesDirectory)/js/react_native/android/gradlew'
-          workingDirectory: '$(Build.SourcesDirectory)/js/react_native/android'
-          options: '--stacktrace'
-          tasks: 'connectedDebugAndroidTest'
-          publishJUnitResults: true
-          testResultsFiles: '**/TEST-*.xml'
-          testRunTitle: 'React Native Android Instrumented Test results'
-          javaHomeOption: 'path'
-          jdkDirectory: '$(JAVA_HOME_11_X64)'
-          sonarQubeRunAnalysis: false
-          spotBugsAnalysis: false
-        displayName: Run React Native Android Instrumented Tests
-
-  - script: |
-      pod install
-    workingDirectory: '$(Build.SourcesDirectory)/js/react_native/ios'
-    displayName: Pod install for onnxruntime react native ios bridge library
-
-  - task: Xcode@5
-    inputs:
-      actions: 'test'
-      configuration: 'Debug'
-      sdk: 'iphonesimulator'
-      xcWorkspacePath: '$(Build.SourcesDirectory)/js/react_native/ios/OnnxruntimeModule.xcworkspace'
-      scheme: 'OnnxruntimeModuleTest'
-      packageApp: false
-      destinationPlatformOption: 'iOS'
-      destinationSimulators: 'iPhone 13,OS=latest'
+- stage: ReactNative_CI
+  displayName: React Native CI
+  dependsOn:
+  - Build_Android_AAR
+  jobs:
+  - job: ReactNative_CI
+    pool:
+      vmImage: 'macOS-12'
+    variables:
+      runCodesignValidationInjection: false
+      TODAY: $[format('{0:dd}{0:MM}{0:yyyy}', pipeline.startTime)]
+      PROTO_CACHE_DIR: '$(Pipeline.Workspace)/ccache_proto'
+      ORT_CACHE_DIR: '$(Pipeline.Workspace)/ccache_ort'
+    timeoutInMinutes: 150
+    steps:
+    - task: mspremier.PostBuildCleanup.PostBuildCleanup-task.PostBuildCleanup@3
+      displayName: Clean Agent Directories
+      condition: always()
+
+    # Onnx has no 3.9 python package available yet, need to use python 3.8 to avoid build onnx package
+    # pythonVersion can be updated in Azure pipeline settings
+    # https://dev.azure.com/onnxruntime/onnxruntime/_build?definitionId=188
+    - task: UsePythonVersion@0
+      displayName: Use Python $(pythonVersion)
+      inputs:
+        versionSpec: $(pythonVersion)
+
+    - task: NodeTool@0
+      inputs:
+        versionSpec: '16.x'
+
+    - script:
+        brew install coreutils ninja npm yarn
+      displayName: Install coreutils, ninja, npm, and yarn
+
+    - script: |
+        python3 -m pip install -q flatbuffers
+      displayName: Install python modules
+
+    - template: mac-build-step-with-cache.yml
+      parameters:
+        TODAY: $(TODAY)
+        AdditionalKey: ' protobuf | "$(Agent.OS)" | $(Build.SourcesDirectory)/cmake/deps.txt, $(Build.SourcesDirectory)/tools/ci_build/github/linux/docker/inference/x64/python/cpu/scripts/install_protobuf.sh'
+        CACHE_DIR: $(PROTO_CACHE_DIR)
+        ChangeEveryCommit: false
```

Comment:
I'm not too familiar with ccache. could you explain what is being cached in this step and the next one, and why this one has ChangeEveryCommit set to false?

Reply:
It‘s used to save storage cost. 
For pipeline cache, new cache is generated with the new key.
Protobuf isn‘t changed frequently and we can check the change by 2 files.  It's a waste to generate new cache in every commit.
Onnxruntime code is updated in every commit, so the Build.SouceVersion(commitid) is part of the key.

Reply:
thanks for the explanation and link.
the protobuf caching makes sense.
for the onnxruntime build, why do we want to have a cache entry for each commit? how often do we expect to build at the same commit?

Reply:
It's too complicated to know which commit would generate different C/C++ outputs.
besides changes in c/c++ files, other files could change compiler options.
`a cache entry for each commit` is to used generate cache for next builds and it's a safe way to keep the hit rate.

With restore keys, the pipeline would download cache generated by previous runs even they aren't same commit.
`This is useful if the pipeline is unable to find an exact match but wants to use a partial cache hit instead.`
https://learn.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#restore-keys.

Other useful info could be checked out in [cache isolation and security]
(https://learn.microsoft.com/en-us/azure/devops/pipelines/release/caching?view=azure-devops#cache-isolation-and-security)
To save storage cost,  I'm trying to force merge branch to read the cache only generated in main branch in #15330

Reply:
I see, thanks.